src/zone.c

```diff
@@ -0,0 +1,191 @@
+{
+	int fd;
+
+	char procfile[MAX_PROCFS_PATH];
+	(void)snprintf(procfile, MAX_PROCFS_PATH, "/proc/%s/%s", pidstr, file);
+	while ((fd = open(procfile, O_RDONLY)) == -1) {
+		if ((errno != EMFILE) || (errno != ENFILE)) {
+			return(1);
+		}
+	}
+
+	if (pread(fd, buf, bufsize, 0) != bufsize) {
+		close(fd);
+		return (1);
+	}
+	close(fd);
+	return (0);
+}
+
+static int
+zone_submit_value(char *zone, gauge_t value)
+{
+	value_list_t vl = VALUE_LIST_INIT;
+	value_t      values[1];
+
+	values[0].gauge = value;
+
+	vl.values = values;
+	vl.values_len = 1; /*STATIC_ARRAY_SIZE (values);*/
+	sstrncpy (vl.host, hostname_g, sizeof (vl.host));
+	sstrncpy (vl.plugin, "zone", sizeof (vl.plugin));
+ 	sstrncpy (vl.type, "zone_cpu", sizeof (vl.type));
+ 	sstrncpy (vl.type_instance, zone, sizeof (vl.type_instance));
+
+	return(plugin_dispatch_values (&vl));
+}
+
+static zone_stats_t *
+zone_find_stats(c_avl_tree_t *tree, zoneid_t zoneid)
+{
+	zone_stats_t *ret = NULL;
+	zoneid_t     *key = NULL;
+
+	if (c_avl_get(tree, (void **)&zoneid, (void **)&ret)) {
+		if (!(ret = malloc(sizeof(zone_stats_t)))) {
+			WARNING("no memory");
+			return(NULL);
+		}
+		if (!(key = malloc(sizeof(zoneid_t)))) {
+			WARNING("no memory");
+			return(NULL);
+		}
+		*key = zoneid;
+		if (c_avl_insert(tree, key, ret)) {
+			WARNING("error inserting into tree");
+			return(NULL);
+		}
+	}
+	return(ret);
+}
+
+static void
+zone_submit_values(c_avl_tree_t *tree)
+{
+	char          zonename[MAX_ZONE_NAME];
+	zoneid_t     *zoneid = NULL;
+	zone_stats_t *stats  = NULL;
+
+	while (c_avl_pick (tree, (void **)&zoneid, (void **)&stats) == 0)
+	{
+		getzonenamebyid(*zoneid, zonename, MAX_ZONE_NAME-1);
+		zone_submit_value(zonename, (gauge_t)FRC2PCT(stats->pctcpu));
+		free(stats);
+		free(zoneid);
+	}
+	c_avl_destroy(tree);
+}
+
+static c_avl_tree_t *
+zone_scandir(DIR *procdir)
+{
+	char         *pidstr;
+	pid_t         pid;
+	dirent_t     *direntp;
+	psinfo_t      psinfo;
+	c_avl_tree_t *tree;
+	zone_stats_t *stats;
+/*	size_t    physmem = sysconf(_SC_PHYS_PAGES) * pagesize;*/
+
+	if (!(tree=c_avl_create((int (*)
+				 (const void *, const void *))zone_compare))) {
+		WARNING("Failed to create tree");
+		return(NULL);
+	}
+
+	for (rewinddir(procdir); (direntp = readdir(procdir)); ) {
+		pidstr = direntp->d_name;
+		if (pidstr[0] == '.')	/* skip "." and ".."  */
+			continue;
+		pid = atoi(pidstr);
+		if (pid == 0 || pid == 2 || pid == 3)
+			continue;	/* skip sched, pageout and fsflush */
+		if (zone_read_procfile(pidstr, "psinfo", &psinfo, 
+				  sizeof(psinfo_t)) != 0)
+			continue;
+		stats = zone_find_stats(tree, psinfo.pr_zoneid);
+		stats->pctcpu += psinfo.pr_pctcpu;
+		stats->pctmem += psinfo.pr_pctmem;
+	}
+	return(tree);
+}
+
+
+static int zone_read (void)
+{
+	DIR          *procdir;
+	c_avl_tree_t *tree;
+
+	pagesize = sysconf(_SC_PAGESIZE);
+	if ((procdir = opendir("/proc")) == NULL) {
+		ERROR("cannot open /proc directory\n");
+		exit(1);
```

Comment:
`return (-1);` – if this function returns non-zero it will be called with an exponential backoff, so the logs won't get flooded.